### PR TITLE
Shadow blocks for context menu

### PR DESCRIPTION
**Pull request by mikeharv**
This is PR is part of the work to transition to new mainline Google Blockly and is a direct follow-up to https://github.com/code-dot-org/code-dot-org/pull/45223

For this work, we are want to allow levelbuilders to toggle the `shadow` block type when building toolboxes and start code in our workspace editors. This allows levelbuilders to make use of Google Blockly's [shadow blocks](https://developers.google.com/blockly/guides/configure/web/toolbox?hl=en#shadow_blocks) feature.

### Crawl
 Make it possible to render shadow blocks and utilize other v7 features, such as `movable` blocks. (https://github.com/code-dot-org/code-dot-org/pull/45200)
### Walk
Update currently published levels to use new Blockly features (e.g. shadow blocks) over cdoBlockly features we want to deprecate (e.g. `can_disconnect_from_parent`). (https://github.com/code-dot-org/code-dot-org/pull/45223) *This branch includes those changes.*
### Run:
Enable levelbuilders to work create and manage levels with the new features. (This PR)
We can now update the block configuration menu that is available in levelbuilder. 

### Run Faster?:
To go even further, we could update our block model to specify a standard and consistent shadow block for a given value field, or work with the curriculum team to design new blocks using the features of Mainline Google Blockly.

## Context and Constraints
Shadow blocks may not include a variable field or have children that are not also shadows. Shadow blocks must have a parent block. Shadow blocks cannot be selected or right-clicked (ie. open a context menu).

For any block that _can_ be a shadow, levelbuilders should be able to make it into one. Because a shadow cannot be right-clicked, the parent block can be used to change shadows back into blocks.

Because of the complexities above, we can use two separate menu options to either A) change a block into a shadow or B) change the shadow children of a block into normal blocks. Each of these menu options should only be presented under the right circumstances.

To summarize, a block can be made into a shadow if:
* It does not contain a variable field.
* It does not have any non-shadow children.
* It has a parent.
A block that has at least one child shadow block will have an option to disable any shadow(s) it contains.

**Standard use:**

https://user-images.githubusercontent.com/43474485/157709625-26bf95b2-9b40-4908-a775-3faff7ead5de.mp4

**Edge case testing:**

https://user-images.githubusercontent.com/43474485/157709659-532e38ea-8037-439f-8fcf-8088d4ba6493.mp4

## Links

- jira ticket: [STAR-2155: [Google Blockly] Add shadow block option to context menu](https://codedotorg.atlassian.net/browse/STAR-2155)

### CODE REVIEW

File: apps/src/blockly/addons/cdoBlockSvg.js [hunk 1]
```diff
@@ -70,20 +70,52 @@ export default class BlockSvg extends GoogleBlockly.BlockSvg {
...
+        const shadow = {
+          text: 'Make Shadow',
+          enabled: true,
+          callback: function() {
+            this.setShadow(true);
+            Blockly.ContextMenu.hide();
+          }.bind(this)
+        };
+        menuOptions.push(shadow);
+      }
```
Comment (katleiahramos) [inline apps/src/blockly/addons/cdoBlockSvg.js]: Ideally I think we should move the logic for determining if a block's context menu should have the shadow option out of this method.

so it would read something like:

```
if (shouldHaveShadowOption()) {
  menuOptions.push(getShadowBlockOption());
 }
 ```

Comment (katleiahramos) [reply apps/src/blockly/addons/cdoBlockSvg.js]: oh actually I didn't realize the other options were declared in this one function as well. so maybe ok to keep  `const shadow = ...` here🤷🏽‍♀️

File: apps/src/blockly/addons/cdoBlockSvg.js [hunk 2]
```diff
@@ -70,20 +70,52 @@ export default class BlockSvg extends GoogleBlockly.BlockSvg {
...
+            text: `Make ${
+              shadowChildCount > 1 ? `${shadowChildCount} ` : ''
+            }Child Block${shadowChildCount > 1 ? 's' : ''} Non-Shadow`,
+            enabled: true,
+            callback: function() {
+              for (let i = 0; i < children.length; i++) {
+                children[i].setShadow(false);
+              }
+              Blockly.ContextMenu.hide();
+            }.bind(this)
```
Comment (katleiahramos) [inline apps/src/blockly/addons/cdoBlockSvg.js]: similarly here, there could be a method called `hasShadowChildren()` that returns a boolean based on if `children.length && shadowChildCount > 0`  or just using that expression to simplify the `if` statement.

**Comment by mikeharv** [comment]
@katleiahramos, I took your suggestions to remove most of the logic from the context menu method. It follows the patterns in this file more closely now and makes things easier to read.

**Comment by katleiahramos** [comment]
hm I may be doing something wrong here, but I can't "unshadow" the behavior (avoiding targets) shadow block in this start mode editor:

image

**Review by katleiahramos — APPROVED** [review]
